packages/context/src/value-promise.ts

```diff
@@ -0,0 +1,160 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * This module contains types for values and/or promises as well as a set of
+ * utility methods to handle values and/or promises.
+ */
+
+// tslint:disable-next-line:no-any
+export type BoundValue = any;
+
+/**
+ * Representing a value or promise. This type is used to represent results of
+ * synchronous/asynchronous resolution of values.
+ */
+export type ValueOrPromise<T> = T | Promise<T>;
+
+export type MapObject<T> = {[name: string]: T};
+
+/**
+ * Check whether a value is a Promise-like instance.
+ * Recognizes both native promises and third-party promise libraries.
+ *
+ * @param value The value to check.
+ */
+export function isPromise<T>(
+  value: T | PromiseLike<T>,
+): value is PromiseLike<T> {
+  if (!value) return false;
+  if (typeof value !== 'object' && typeof value !== 'function') return false;
+  return typeof (value as PromiseLike<T>).then === 'function';
+}
+
+/**
+ * Get nested properties by path
+ * @param value Value of an object
+ * @param path Path to the property
+ */
+export function getDeepProperty(value: BoundValue, path: string) {
+  const props = path.split('.').filter(Boolean);
+  for (const p of props) {
+    if (value == null) {
+      return value;
+    }
+    value = value[p];
+  }
+  return value;
+}
+
+/**
+ * Resolve entries of an object into a new object with the same keys or a
+ * promise of the new object. Please note some of the entries can be resolved
+ * synchronously while others need to be resolved asynchronously.
```

Comment:
> Resolve entries of an object into a new object with the same keys or a promise of the new object.

I find this part difficult to parse and understand:

> a new object with the same keys or a promise of the new object.

I am proposing the following text:

> Resolve entries of an object into a new object with the same keys. If any of the entries require asynchronous resolution, then this function returns a promise of the new object instead. The promise will be resolved with an object containing the resolved values, i.e. any promises returned by individual resolvers are handled by this function and the resulting object is promise-free.

The same comment applies to the docs for `resolveList` too.